compiler/src/dotty/tools/dotc/ast/MainProxies.scala

```diff
@@ -115,4 +136,256 @@ object MainProxies {
+              // Arguments to be passed to ParameterInfos' constructor
+              List(List(lit(param), lit(formalType.show)))
+            )
+
+            /*
+             * Assignations to be made after the creation of the ParameterInfos.
+             * For example:
+             *   args0paramInfos.withDocumentation("my param x")
+             * is represented by the pair
+             *   (defn.MainAnnotationParameterInfos_withDocumentation, List(lit("my param x")))
+             */
+            var assignations: List[(Symbol, List[Tree])] = Nil
+            for (dvSym <- defaultValueSymbols.get(n))
+              assignations = (defn.MainAnnotationParameterInfos_withDefaultValue -> List(unitToValue(ref(dvSym.termRef)))) :: assignations
+            for (doc <- documentation.argDocs.get(param))
+              assignations = (defn.MainAnnotationParameterInfos_withDocumentation -> List(lit(doc))) :: assignations
+
+            val instanciatedAnnots = paramAnnotations(n).map(instanciateAnnotation).toList
+            if instanciatedAnnots.nonEmpty then
+              assignations = (defn.MainAnnotationParameterInfos_withAnnotations -> instanciatedAnnots) :: assignations
+
+            if assignations.isEmpty then
+              paramInfosTree
+            else
+              assignations.foldLeft[Tree](paramInfosTree){ case (tree, (setterSym, values)) => Apply(Select(tree, setterSym.name), values) }
+          }
+
+          val argDef = ValDef(
+            argName,
+            TypeTree(),
+            Apply(TypeApply(Select(Ident(cmdName), getterSym.name), TypeTree(formalType) :: Nil), parameterInfos),
+          )
+
+          (argRef, argDef)
+      }
+    end createArgs
+
+    /** Turns an annotation (e.g. `@main(40)`) into an instance of the class (e.g. `new scala.main(40)`). */
+    def instanciateAnnotation(annot: Annotation): Tree =
+      val argss = {
+        def recurse(t: tpd.Tree, acc: List[List[Tree]]): List[List[Tree]] = t match {
+          case Apply(t, args: List[tpd.Tree]) => recurse(t, extractArgs(args) :: acc)
+          case _ => acc
+        }
+
+        def extractArgs(args: List[tpd.Tree]): List[Tree] =
+          args.flatMap {
+            case Typed(SeqLiteral(varargs, _), _) => varargs.map(arg => TypedSplice(arg))
+            case arg: Select if arg.name.is(DefaultGetterName) => Nil  // Ignore default values, they will be added later by the compiler
+            case arg => List(TypedSplice(arg))
+          }
+
+        recurse(annot.tree, Nil)
+      }
+
+      New(TypeTree(annot.symbol.typeRef), argss)
+    end instanciateAnnotation
+
+    var result: List[TypeDef] = Nil
+    if (!mainFun.owner.isStaticOwner)
+      report.error(s"main method is not statically accessible", pos)
+    else {
+      val cmd = ValDef(
+        cmdName,
+        TypeTree(),
+        Apply(
+          Select(instanciateAnnotation(mainAnnot), defn.MainAnnot_command.name),
```

Comment:
```suggestion
          Select(instanciateAnnotation(mainAnnot), nme.main),
```

Reply:
That won't work, `nme.main` is "scala.main" whereas `defn.MainAnnot_command` is "scala.main.command"

Reply:
Then we should add `command` to `StdNames`.